### PR TITLE
WIP: Json workflow config API endpoint

### PR DESCRIPTION
**Pull request by axelri**
Implement a new API endpoint for letting users create and update workflows by uploading a json configuration payload.

### CODE REVIEW

**Comment by axelri** [comment]
PTAL @honnix @ulzha

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 1]
```diff
@@ -486,6 +504,103 @@ public void shouldPaginateWorkflowInstancesData() throws Exception {
...
+    assertThat(response, hasNoPayload());
+  }
+
+  @Test
+  public void shouldReturnInternalServerErrorWhenSchedulerRequestFails() throws Exception {
+    sinceVersion(Api.Version.V3);
+
+    Response<ByteString> response =
+        awaitResponse(
+            serviceHelper.request("POST", path("/foo"), serialize(WORKFLOW_CONFIGURATION)));
```
Comment (ulzha) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: Had a little trouble readability-wise (what is it that ensures that scheduler request fails in this test). Suggestion: path("/intentionally-nonexistent-path") or something like that this time, instead of path("/foo")  as in the above tests, where it's really a "foo" (irrelevant) value.

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 2]
```diff
@@ -486,6 +504,103 @@ public void shouldPaginateWorkflowInstancesData() throws Exception {
...
+        awaitResponse(
+            serviceHelper.request("DELETE", path("/foo/bar")));
+
+    assertThat(response, hasStatus(withCode(Status.INTERNAL_SERVER_ERROR)));
+    assertThat(response.status(), withReasonPhrase(equalTo("Error in call to Scheduler")));
+    assertThat(response, hasNoPayload());
+  }
+
+  @Test
+  public void shouldForwardInteralResponseForDeleteWorkflow() throws Exception {
```
Comment (ulzha) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: Typo "Interal"

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 3]
```diff
@@ -486,6 +504,103 @@ public void shouldPaginateWorkflowInstancesData() throws Exception {
...
+    assertThat(response, hasStatus(withCode(Status.OK)));
+    assertThat(deserialize(response.payload().get(), Workflow.class), equalTo(WORKFLOW2));
+  }
+
+  @Test
+  public void shouldReturnErrorMessageWhenSchedulerFailsWorkflow() throws Exception {
+    sinceVersion(Api.Version.V3);
+    serviceHelper.stubClient()
+        .respond(Response.forStatus(Status.SERVICE_UNAVAILABLE))
+        .to(SCHEDULER_BASE + "/api/v0/workflows");
```
Comment (ulzha) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: Is this stub invoked in this test at all? The `to()` uri doesn't match "/foo" and AFAIU they're matched strictly?

Comment (axelri) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: Yes, the stubClient is invoked within the WorkflowResource context to stub the internal API calls to SchedulerResource. Hence, the URIs does not have to match.

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java (hunk 2 above)
Comment (axelri) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: Fixed https://github.com/spotify/styx/pull/243/commits/08e7f13796a28c831f3e20f5b436cf9276d4614a

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java (hunk 1 above)
Comment (axelri) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: Added a comment to clarify this test. The stub client is intentionally not setup, this mimics a unreachable internal service (SchedulerResource). The path of the original call should still be /foo, because it corresponds to the component id (/<cid>) and should be a valid path in WorkflowResource  https://github.com/spotify/styx/pull/243/commits/08e7f13796a28c831f3e20f5b436cf9276d4614a

File: doc/api.apib [hunk 4]
```diff
@@ -53,6 +53,67 @@ Resources related to the Workflows registered in Styx.
...
+
+### Create or update Workflow [POST]
+
++ Request (application/json)
+
+        {
+          "id": "StyxCanary",
+          "schedule": "days",
+          "offset": null,
+          "docker_image": null,
```
Comment (ulzha) [inline doc/api.apib]: Suggest `"styx-canary-dummy:dummy"`? Part of the benefit that I hope we can reap from this endpoint is  that we would completely disallow missing images.

Comment (axelri) [inline doc/api.apib]: Updated the API to only accept non-null values for docker_image https://github.com/spotify/styx/pull/243/commits/7e068b2e79dcbd847051cd864794f176768bb72c

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 5]
```diff
@@ -77,12 +92,80 @@ public WorkflowResource(Storage storage) {
...
+        )
     );
 
     return Api.prefixRoutes(routes, V2, V3);
   }
 
+  private Response<ByteString> deleteWorkflow(String cid, String wfid, RequestContext rc) {
+    Request request = Request.forUri(schedulerApiUrl("workflows", cid, wfid), "DELETE");
+    try {
+      return rc.requestScopedClient().send(request).toCompletableFuture().get();
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: Better not to block explicitly in user code, instead letting Apollo handle it. You can find some examples in `BackfillResource`.

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 6]
```diff
@@ -77,12 +92,80 @@ public WorkflowResource(Storage storage) {
...
+      requestPayload = serialize(workflow);
+    } catch (JsonProcessingException e) {
+      return Response.forStatus(
+          Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Failed to serialize proxy payload."));
+    }
+
+    final Request request = Request.forUri(schedulerApiUrl("workflows"), "POST")
+        .withPayload(requestPayload);
+    try {
+      StatusType responseStatus = rc.requestScopedClient().send(request).toCompletableFuture().get()
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: Same here, not to block.

File: styx-api-service/src/main/java/com/spotify/styx/StyxApi.java [hunk 7]
```diff
@@ -66,6 +66,7 @@
 
   public static final String SCHEDULER_SERVICE_BASE_URL = "styx.scheduler.base-url";
   public static final String DEFAULT_SCHEDULER_SERVICE_BASE_URL = "http://localhost:8080";
+  public static final String COMPONENTS_API_BASE_URL = "styx.components.api.base-url";
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/StyxApi.java]: Let's just get rid of this field and leave it empty when posting to API endpoint.

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 8]
```diff
@@ -99,9 +118,9 @@ public WorkflowResourceTest(Api.Version version) {
 
   private static final ByteString STATEPAYLOAD_FULL =
       ByteString.encodeUtf8("{\"enabled\":\"true\", \"docker_image\":\"cherry:image\", "
-                            + "\"commit_sha\":\"" + VALID_SHA
-                            + "\", \"next_natural_trigger\":\"2016-08-10T07:00:01Z\", "
-                            + "\"next_natural_offset_trigger\":\"2016-08-10T08:00:01Z\"}");
+          + "\"commit_sha\":\"" + VALID_SHA
```
Comment (honnix) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: This seems not aligned with default formatter. Also applies to other format related changes.

File: styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java [hunk 9]
```diff
@@ -148,4 +189,9 @@ public SchedulerResource(
     // todo: change payload to a struct returning the triggerId as well so the user can refer to it
     return Response.forPayload(workflowInstance);
   }
+
+  private static String arg(String name, RequestContext rc) {
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java]: I think you can just inline this as there is only one place calling this method.

File: styx-scheduler-service/src/test/java/com/spotify/styx/api/SchedulerResourceTest.java [hunk 10]
```diff
@@ -66,27 +67,44 @@
  */
 public class SchedulerResourceTest {
 
-  private final InMemStorage storage = new InMemStorage();
+  private InMemStorage storage = new InMemStorage();
```
Comment (honnix) [inline styx-scheduler-service/src/test/java/com/spotify/styx/api/SchedulerResourceTest.java]: Not `final` anymore?

File: styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java [hunk 11]
```diff
@@ -82,11 +92,42 @@ public SchedulerResource(
...
+    )
 
         .map(r -> r.withMiddleware(Middleware::syncToAsync));
   }
 
+  private Response<ByteString> deleteWorkflow(String cid, String wfid) {
+    WorkflowId workflowId = WorkflowId.create(cid, wfid);
+    Optional<Workflow> workflowOpt = null;
+    try {
+      workflowOpt = storage.workflow(workflowId);
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java]: existence check can be done already in API service to avoid proxying unnecessary requests.

**Comment by honnix** [comment]
In general this PR is good. Please fix the formatting issue first, otherwise it's hard to follow all the changes.

File: styx-api-service/src/main/java/com/spotify/styx/StyxApi.java (hunk 7 above)
Comment (axelri) [inline styx-api-service/src/main/java/com/spotify/styx/StyxApi.java]: https://github.com/spotify/styx/pull/243/commits/b6ab4fcf0e612735224f3fa5402205f75ceb2a53

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 5 above)
Comment (axelri) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: thanks, fixed https://github.com/spotify/styx/pull/243/commits/849afae4f079c5bd65f86b985642da04b2d977d2

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 6 above)
Comment (axelri) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: https://github.com/spotify/styx/pull/243/commits/849afae4f079c5bd65f86b985642da04b2d977d2

File: styx-scheduler-service/src/test/java/com/spotify/styx/api/SchedulerResourceTest.java (hunk 10 above)
Comment (axelri) [inline styx-scheduler-service/src/test/java/com/spotify/styx/api/SchedulerResourceTest.java]: That's because we assign new storage object before every test, since our tests depend on a clean storage object. Arguably it would be better to implement a clean() method in the InMemoryStorage class.

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 12]
```diff
@@ -92,78 +87,68 @@ public WorkflowResource(Storage storage, String schedulerServiceBaseUrl,
...
+  private CompletionStage<Response<ByteString>> deleteWorkflow(String cid, String wfid,
+                                                               RequestContext rc) {
     Request request = Request.forUri(schedulerApiUrl("workflows", cid, wfid), "DELETE");
-    try {
-      return rc.requestScopedClient().send(request).toCompletableFuture().get();
-    } catch (InterruptedException | ExecutionException e) {
-      return Response
-          .forStatus(Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Error in call to Scheduler"));
-    }
+    return rc.requestScopedClient().send(request);
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: Something like `rc.requestScopedClient().send(rc.request().withUri(schedulerApiUrl("workflows", cid, wfid)));` instead to keep additional header, etc.

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 13]
```diff
@@ -92,78 +87,68 @@ public WorkflowResource(Storage storage, String schedulerServiceBaseUrl,
...
-      if (responseStatus.family().equals(SUCCESSFUL)) {
-        return Response.forPayload(workflow);
-      } else {
-        return Response.forStatus(responseStatus);
-      }
-    } catch (ExecutionException | InterruptedException e) {
-      return Response
-          .forStatus(Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Error in call to Scheduler"));
-    }
+    return rc.requestScopedClient().send(request);
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: Same here to keep additional headers, etc.

File: styx-common/src/main/java/com/spotify/styx/model/Workflow.java [hunk 14]
```diff
@@ -37,7 +38,7 @@
   public abstract String workflowId();
 
   @JsonProperty
-  public abstract URI componentUri();
+  public abstract Optional<URI> componentUri();
```
Comment (ulzha) [inline styx-common/src/main/java/com/spotify/styx/model/Workflow.java]: Maybe add a slight bit of documentation to this? Even if of the "legacy, don't use" sort.

File: styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java (hunk 11 above)
Comment (axelri) [inline styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java]: We discussed this and decided to do all logic in the SchedulerResource, while WorkflowResource simply functions as a proxy

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 12 above)
Comment (axelri) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: https://github.com/spotify/styx/pull/243/commits/ddc08bbbbbbf08772b2dddb7ce832f3deb99f8bf

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 13 above)
Comment (axelri) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: https://github.com/spotify/styx/pull/243/commits/ddc08bbbbbbf08772b2dddb7ce832f3deb99f8bf

File: styx-common/src/main/java/com/spotify/styx/model/Workflow.java (hunk 14 above)
Comment (axelri) [inline styx-common/src/main/java/com/spotify/styx/model/Workflow.java]: Added some short comments https://github.com/spotify/styx/pull/243/commits/f823aacb73c7de162e1eadaffebf7dba07ac88c6

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 15]
```diff
@@ -146,9 +145,7 @@ public WorkflowResource(Storage storage, String schedulerServiceBaseUrl) {
           Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Failed to serialize proxy payload.")));
     }
 
-    final Request request = Request.forUri(schedulerApiUrl("workflows"), "POST")
-        .withPayload(requestPayload);
-    return rc.requestScopedClient().send(request);
+    return rc.requestScopedClient().send(rc.request().withUri(schedulerApiUrl("workflows")));
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: Payload?

**Comment by honnix** [comment]
It would be helpful to squash some commits. Getting hard to review.

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 16]
```diff
@@ -78,9 +88,65 @@ public WorkflowResource(Storage storage) {
...
+  }
+
+  private CompletionStage<Response<ByteString>> createOrUpdateWorkflow(String componentId,
+                                                                       RequestContext rc) {
+    final Optional<ByteString> payload = rc.request().payload();
+    if (!payload.isPresent()) {
+      return CompletableFuture.completedFuture(
+          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Missing payload.")));
+    }
+    WorkflowConfiguration workflowConfig;
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: `final`

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 17]
```diff
@@ -78,9 +88,65 @@ public WorkflowResource(Storage storage) {
...
+      return CompletableFuture.completedFuture(
+          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Invalid payload.")));
+    }
+
+    if (!workflowConfig.dockerImage().isPresent()) {
+      return CompletableFuture.completedFuture(
+          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("docker_image is required")));
+    }
+
+    Workflow workflow = Workflow.create(componentId, Optional.empty(), workflowConfig);
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: `final`

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 18]
```diff
@@ -213,6 +279,11 @@ public WorkflowResource(Storage storage) {
     }
   }
 
+  // TODO: duplicate code
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: I think it is fine to duplicate a bit.

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 19]
```diff
@@ -177,7 +197,7 @@ public void shouldSucceedWithFullPatchStatePerWorkflow() throws Exception {
 
     response =
         awaitResponse(serviceHelper.request("PATCH", path("/foo/bar/state"),
-                                            STATEPAYLOAD_FULL));
+            STATEPAYLOAD_FULL));
```
Comment (honnix) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java]: This is still not well-formatted.

**Comment by honnix** [comment]
👍
